kernel/src/hil/flash.rs

```diff
@@ -142,3 +164,111 @@ pub trait Client<F: Flash> {
     /// Flash erase complete.
     fn erase_complete(&self, error: Error);
 }
+
+/// A checked flash region
+pub struct Page {
+    pub address: u64,
+    pub length: u32,
+}
+
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
+    /// `address` is calculated as an offset from the start of the flash
+    /// region.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    /// If `ErrorCode::NOSUPPORT` is returned then `read_page()`
+    // should be used instead.
+    fn read(
+        &self,
+        region: &Page,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Create a `Page` from the provided `address` and `length`.
+    ///
+    /// The `Page->address` and Page->length` will be modified to match
```

Comment:
There should be an elaboration on what modification exactly happens, e.g.:
```suggestion
    /// The resulting `Page` will cover the requested area, adjusting the provided values according to
```